kore/src/Kore/Step/SMT/Representation/Symbols.hs

```diff
@@ -71,8 +71,8 @@ buildRepresentations
     -> AST.UnresolvedDeclarations
 buildRepresentations indexedModule =
     listToDeclarations builtinDeclarations
-    `AST.mergePreferFirst` listToDeclarations constructorDeclarations
     `AST.mergePreferFirst` listToDeclarations smtlibDeclarations
+    `AST.mergePreferFirst` listToDeclarations constructorDeclarations
```

Comment:
What is the effect of moving the `constructorDeclarations` to the end? I would have expected that we _prefer_ sending constructor declarations to the solver.

Reply:
If I can remember correctly, @virgil-serbanuta explained that swapping these two lines would result in the symbol declaration being represented as https://github.com/kframework/kore/blob/d13e1a0c3257f1e2e52764468b5e3a021832b8f1/kore/src/Kore/Step/SMT/AST.hs#L203 instead of as https://github.com/kframework/kore/blob/d13e1a0c3257f1e2e52764468b5e3a021832b8f1/kore/src/SMT/AST.hs#L179. I hope this is correct, if not, I'm sure @virgil-serbanuta has a much better idea of what is going on here.

Reply:
Well, the current change moves the constructor declarations in the middle, precisely because that makes sure that we don't send smtlib declarations for constructors.

Constructors are declared when we declare the sort, the entry in `constructorDeclarations` is used only for bookkeeping, since we need to know what symbols does the SMT understand. The mechanism through which that happens is, indeed the `IndirectSymbolDeclaration` vs `FunctionDeclaration` thing that Ana mentioned.